sympy/matrices/matrices.py

```diff
@@ -2363,10 +2363,13 @@ def _matrix_pow_by_jordan_blocks(self, num):
         def jordan_cell_power(jc, n):
             N = jc.shape[0]
             l = jc[0, 0]
-            if l == 0 and (n < N - 1) != False:
-                raise ValueError("Matrix det == 0; not invertible")
-            elif l == 0 and N > 1 and n % 1 != 0:
-                raise ValueError("Non-integer power cannot be evaluated")
+            if l == 0:
+                if isinstance (n, Expr) and not n.is_extended_real:
+                    raise NonInvertibleMatrixError("Non-invertible matrix to non-real power")
+                if (n < N - 1) != False:
```

Comment:
Maybe this whole block can become the equivalent of:
```python
if l == 0:
    if not (n.is_integer and n.is_nonnegative):
        raise NonInvertibleMatrixError("Non-invertible matrix can only be raised to a nonnegative integer power")
    return Matrix(N, N, lambda i, j: KroneckerDelta(j-i, n) if i<=j else 0)
```

This is an internal function so it shouldn't be necessary to sympify n (that should happen before this is called).

Reply:
Doesn't work, gives:
```
A = Matrix([[0, 1, 0], [0, 0, 1], [0, 0, 1]])
n = Symbol('n', integer=True, nonnegative=True)
A**n == A
```

Reply:
I get this:
```julia
In [106]: M                                                                                                                       
Out[106]: 
⎡δ     δ     - δ    - δ    + 1⎤
⎢ 0,n   1,n     0,n    1,n    ⎥
⎢                             ⎥
⎢ 0    δ         1 - δ        ⎥
⎢       0,n           0,n     ⎥
⎢                             ⎥
⎣ 0     0            1        ⎦

In [107]: M.subs(n, 1)                                                                                                            
Out[107]: 
⎡0  1  0⎤
⎢       ⎥
⎢0  0  1⎥
⎢       ⎥
⎣0  0  1⎦

In [108]: M.subs(n, 0)                                                                                                            
Out[108]: 
⎡1  0  0⎤
⎢       ⎥
⎢0  1  0⎥
⎢       ⎥
⎣0  0  1⎦

In [109]: M.subs(n, 2)                                                                                                            
Out[109]: 
⎡0  0  1⎤
⎢       ⎥
⎢0  0  1⎥
⎢       ⎥
⎣0  0  1⎦

In [110]: M.subs(n, 3)                                                                                                            
Out[110]: 
⎡0  0  1⎤
⎢       ⎥
⎢0  0  1⎥
⎢       ⎥
⎣0  0  1⎦
```
Note when I said the "equivalent of" I meant that you can't literally use return here because the matrix should be modified in place (according to how the function is set up now, although I'd be tempted to rewrite that).

Reply:
Ok sorry, working on two things at the same time. Your fix works and can be plugged in changing only one test case, but why was the original function failing on that input, that could be another hidden bug?

Reply:
I assume the failing input you ask about is `Matrix([[0, 1, 0], [0, 0, 1], [0, 0, 1]])**n`. The original function was failing because of the line quoted above these messages:
```
                if (n < N - 1) != False:
```
Here n is a symbol so this inequality doesn't evaluate.

Where n and N are concrete integers this test was needed before to stop `l**(n-i)*bn` from computing `0**(n-i)` when n-i is negative. However looking at the nilpotent `l=0` case I realised that it has a specific form that is very different from the other cases and doesn't need binomial coefficients etc. Now we compute that form separately, explicitly and symbolically.

> It seems like an extremely specific fix though...

 Definitely a corner case :)

The end result here is that (assuming that the Jordan form is computable) we can calculate any matrix to the power of a symbolic nonnegative integer. I think it's worth a bit of effort to close that gap!

Actually part of the reason I was looking at it was to think about fractional powers of nilpotent Jordan blocks. I can now see that they are poorly defined. Any nilpotent matrix has infinitely many fractional powers. For example if `A**2 = 0` then `A**2 = B**3` for any matrix satisfying `B**3 = 0` so `A**(2/3)` can be basically any nilpotent matrix of index 3 or less. Of course there is a natural choice `B = 0` if we want to define `A**(2/3)`. I'm not sure if that's reasonable to do though.

Reply:
Actually I was just playing with that very line and it was what was causing the original algorithm to break. I think I fixed it, will send it up and you tell me if I am on crack or if this is it...

Reply:
Some tests change because now we get the same effect as your KroneckerDelta fix for other cases. But it does break something:
```
A = Matrix([[0, 1, 0], [0, 0, 1], [0, 0, 1]])
n = Symbol('n', integer=True, nonnegative=True)
A**n
Matrix([
[0**n,    0, 1 - 0**n],
[   0, 0**n, 1 - 0**n],
[   0,    0,        1]])

M = _
M.subs (n, 2)
Matrix([
[0, 0, 1],
[0, 0, 1],
[0, 0, 1]])

M.subs (n, -1)
Matrix([
[zoo,   0, zoo],
[  0, zoo, zoo],
[  0,   0,   1]])
```
Though that last case is what we talked about where you substitute something with different properties than the symbol had on evaluation.

Reply:
It's almost there (see the n=1 case below):
```julia
In [10]: n = Symbol('n', integer=True, nonnegative=True)                                                                          

In [11]: M = Matrix([[0, 1, 0], [0, 0, 1], [0, 0, 1]])                                                                            

In [12]: M**n                                                                                                                     
Out[12]: 
⎡ n           n⎤
⎢0   0   1 - 0 ⎥
⎢              ⎥
⎢     n       n⎥
⎢0   0   1 - 0 ⎥
⎢              ⎥
⎣0   0     1   ⎦

In [13]: (M**n).subs(n, 0)                                                                                                        
Out[13]: 
⎡1  0  0⎤
⎢       ⎥
⎢0  1  0⎥
⎢       ⎥
⎣0  0  1⎦

In [14]: (M**n).subs(n, 1)                                                                                                        
Out[14]: 
⎡0  0  1⎤
⎢       ⎥
⎢0  0  1⎥
⎢       ⎥
⎣0  0  1⎦

In [15]: M                                                                                                                        
Out[15]: 
⎡0  1  0⎤
⎢       ⎥
⎢0  0  1⎥
⎢       ⎥
⎣0  0  1⎦
```

Reply:
(I would prefer not to depend on `0**0=1` for the `n=0` case though because I always find that dubious)

Reply:
Yeah I just saw that too, need to break, will continue on this tomorrow...